src/charm.py

```diff
@@ -0,0 +1,166 @@
+            return False
+        self.unit.status = MaintenanceStatus("Kubernetes authentication completed.")
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/root/.kube/config", "root", self.unit.name
+        )
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/home/ubuntu/.kube/config", "ubuntu", self.unit.name
+        )
+        return True
+
+    def _check_certificates(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self.CA_CERT_PATH.write_text(self.certificates.ca)
+        return True
+
+    def _check_config(self, event: EventBase) -> None:  # TODO: Not the best name
+        if not self._check_certificates(event):
+            return
+
+        if not self._check_kube_control(event):
+            return
+
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+        self.unit.status = ops.ActiveStatus("Ready to test.")
+
+    def _upgrade_charm(self, event: EventBase) -> None:
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+    def _install_snaps(self, channel: Optional[str]) -> None:
+        self.unit.status = ops.MaintenanceStatus("Installing core snap.")
+        self._ensure_snap("core")
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubectl snap.")
+        self._ensure_snap("kubectl", channel=channel)
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubernetes-test snap.")
+        self._ensure_snap("kubernetes-test", channel=channel, classic=True)
+
+    def _ensure_snap(
+        self,
+        name: str,
+        state: SnapState = SnapState.Latest,
+        channel: Optional[str] = "",
+        classic: Optional[bool] = False,
+    ) -> None:
+        if not isinstance(name, str) or name == "":
+            raise ValueError("A name is required to ensure a snap.")
+
+        snap = self.snap_cache[name]
+        if not snap.present:
+            snap.ensure(state=state, classic=classic, channel=channel)
+
+    def _on_start(self, event: EventBase) -> None:
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_test_action(self, event: ActionEvent) -> None:
+        focus = str(event.params.get("focus", ""))
+        parallelism = str(event.params.get("parallelism", ""))
+        skip = str(event.params.get("skip", ""))
+        timeout = str(event.params.get("timeout", ""))
+        extra = str(event.params.get("extra", ""))
+
+        command = ["scripts/test.sh", focus, skip, parallelism, timeout, extra]
```

Comment:
do you like this

```suggestion
        param_get = lambda p: str(event.param.get(p, ""))
        args = [param_get(v) for param in ["focus", "parallelism", "skip", "timeout", "extra"])
        command = ["scripts/test.sh", *args]
```

Reply:
TIL: https://docs.astral.sh/ruff/rules/lambda-assignment/